src/DataTypes/DataTypeObject.cpp

```diff
@@ -1,82 +1,506 @@
+    }
+
+    std::vector<String> sorted_typed_paths;
+    sorted_typed_paths.reserve(typed_paths.size());
+    for (const auto & [path, _] : typed_paths)
+        sorted_typed_paths.push_back(path);
+    std::sort(sorted_typed_paths.begin(), sorted_typed_paths.end());
+    for (const auto & path : sorted_typed_paths)
+    {
+        write_separator();
+        out << backQuoteIfNeed(path) << " " << typed_paths.at(path)->getName();
+    }
+
+    std::vector<String> sorted_skip_paths;
+    sorted_skip_paths.reserve(paths_to_skip.size());
+    for (const auto & skip_path : paths_to_skip)
+        sorted_skip_paths.push_back(skip_path);
+    std::sort(sorted_skip_paths.begin(), sorted_skip_paths.end());
+    for (const auto & skip_path : sorted_skip_paths)
+    {
+        write_separator();
+        out << "SKIP " << backQuoteIfNeed(skip_path);
+    }
+
+    for (const auto & skip_regexp : path_regexps_to_skip)
+    {
+        write_separator();
+        out << "SKIP REGEXP " << quoteString(skip_regexp);
+    }
+
+    if (!first)
+        out << ")";
+
     return out.str();
 }
 
-static DataTypePtr create(const ASTPtr & arguments)
+MutableColumnPtr DataTypeObject::createColumn() const
+{
+    std::unordered_map<String, MutableColumnPtr> typed_path_columns;
+    typed_path_columns.reserve(typed_paths.size());
+    for (const auto & [path, type] : typed_paths)
+        typed_path_columns[path] = type->createColumn();
+
+    return ColumnObject::create(std::move(typed_path_columns), max_dynamic_paths, max_dynamic_types);
+}
+
+namespace
+{
+
+/// It is possible to have nested JSON object inside Dynamic. For example when we have an array of JSON objects.
+/// During type inference in parsing in case of creating nested JSON objects, we reduce max_dynamic_paths/max_dynamic_types by factors
+/// NESTED_OBJECT_MAX_DYNAMIC_PATHS_REDUCE_FACTOR/NESTED_OBJECT_MAX_DYNAMIC_TYPES_REDUCE_FACTOR.
+/// So the type name will actually be JSON(max_dynamic_paths=N, max_dynamic_types=M). But we want the user to be able to query it
+/// using json.array.:`Array(JSON)`.some.path without specifying max_dynamic_paths/max_dynamic_types.
+/// To support it, we do a trick - we replace JSON name in subcolumn to JSON(max_dynamic_paths=N, max_dynamic_types=M), because we know
+/// the exact values of max_dynamic_paths/max_dynamic_types for it.
+void replaceJSONTypeNameIfNeeded(String & type_name, size_t max_dynamic_paths, size_t max_dynamic_types)
+{
+    auto pos = type_name.find("JSON");
+    while (pos != String::npos)
+    {
+        /// Replace only if we don't already have parameters in JSON type declaration.
+        if (pos + 4 == type_name.size() || type_name[pos + 4] != '(')
+            type_name.replace(pos, 4, fmt::format("JSON(max_dynamic_paths={}, max_dynamic_types={})", max_dynamic_paths / DataTypeObject::NESTED_OBJECT_MAX_DYNAMIC_PATHS_REDUCE_FACTOR, std::max(max_dynamic_types / DataTypeObject::NESTED_OBJECT_MAX_DYNAMIC_TYPES_REDUCE_FACTOR, 1lu)));
```

Comment:
> why are max_dynamic_paths protected by max?

Because we couldn't have `max_dynamic_types=0` in old Dynamoc implementation, now after reimplementation we can have it. Will update when the reimplementation is merged.